pkg/store/cache/caching_bucket.go

```diff
@@ -132,35 +131,184 @@ func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailur
+		if err == nil {
+			cb.operationHits.WithLabelValues(opIter, cfgName).Inc()
+			for _, n := range list {
+				if err := f(n); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+		level.Warn(cb.logger).Log("msg", "failed to decode cached Iter result", "err", err)
+	}
+
+	// Iteration can take a while (esp. since it calls function), and iterTTL is generally low.
+	// We will compute TTL based time when iteration started.
+	iterTime := time.Now()
+	var list []string
+	err := cb.Bucket.Iter(ctx, dir, func(s string) error {
+		list = append(list, s)
+		return f(s)
+	})
+
+	remainingTTL := cfg.ttl - time.Since(iterTime)
+	if err == nil && remainingTTL > 0 {
+		if data, err := cfg.codec.Encode(list); err == nil {
+			cfg.cache.Store(ctx, map[string][]byte{key: data}, remainingTTL)
+		} else {
+			level.Warn(cb.logger).Log("msg", "failed to encode Iter result", "err", err)
+		}
+	}
+	return err
+}
+
+func (cb *CachingBucket) Exists(ctx context.Context, name string) (bool, error) {
+	cfgName, cfg := cb.cfg.findExistConfig(name)
+	if cfg == nil {
+		return cb.Bucket.Exists(ctx, name)
+	}
+
+	cb.operationRequests.WithLabelValues(opExists, cfgName).Inc()
+
+	key := cachingKeyExists(name)
+	hits := cfg.cache.Fetch(ctx, []string{key})
+
+	if ex := hits[key]; ex != nil {
+		exists, err := strconv.ParseBool(string(ex))
+		if err == nil {
+			cb.operationHits.WithLabelValues(opExists, cfgName).Inc()
+			return exists, nil
+		}
+		level.Warn(cb.logger).Log("msg", "unexpected cached 'exists' value", "val", string(ex))
+	}
+
+	existsTime := time.Now()
+	ok, err := cb.Bucket.Exists(ctx, name)
+	if err == nil {
+		storeExistsCacheEntry(ctx, key, ok, existsTime, cfg.cache, cfg.existsTTL, cfg.doesntExistTTL)
+	}
+
+	return ok, err
+}
+
+func storeExistsCacheEntry(ctx context.Context, cachingKey string, exists bool, ts time.Time, cache cache.Cache, existsTTL, doesntExistTTL time.Duration) {
+	var ttl time.Duration
+	if exists {
+		ttl = existsTTL - time.Since(ts)
+	} else {
+		ttl = doesntExistTTL - time.Since(ts)
+	}
+
+	if ttl > 0 {
+		cache.Store(ctx, map[string][]byte{cachingKey: []byte(strconv.FormatBool(exists))}, ttl)
+	}
+}
+
+func (cb *CachingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	cfgName, cfg := cb.cfg.findGetConfig(name)
+	if cfg == nil {
+		return cb.Bucket.Get(ctx, name)
+	}
+
+	cb.operationRequests.WithLabelValues(opGet, cfgName).Inc()
+
+	contentKey := cachingKeyContent(name)
+	existsKey := cachingKeyExists(name)
+
+	hits := cfg.cache.Fetch(ctx, []string{contentKey, existsKey})
+	if hits[contentKey] != nil {
+		cb.operationHits.WithLabelValues(opGet, cfgName).Inc()
+		return ioutil.NopCloser(bytes.NewReader(hits[contentKey])), nil
+	}
+
+	// If we know that file doesn't exist, we can return that. Useful for deletion marks.
+	if ex := hits[existsKey]; ex != nil {
+		if exists, err := strconv.ParseBool(string(ex)); err == nil && !exists {
+			cb.operationHits.WithLabelValues(opGet, cfgName).Inc()
```

Comment:
can we store those information in exists cache actually?

Reply:
and share?

Reply:
actually we do lol, but the metrics I would expect to be incremented for exists then,.... same below

Reply:
> but the metrics I would expect to be incremented for exists then,.... same below

I was thinking about this, but I think it would be confusing to report it for `exists` operation... this is really `get` operation, but reusing cached information about whether file exist. I want 'requests' and 'hits' metrics to be comparable... if we increase 'hits' for exists, we also need to increase 'requests' ... but then 'requests' will not match real bucket usage, which is the confusing bit. Personally I think it's ok the way it is.

Reply:
Then I might think we should adjust request for exists as well (: 

But happy to leave it like this and adjust later if needed.